src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenForEachTests.cs

```diff
@@ -2366,12 +2366,661 @@ .locals init (T V_0)
+    IL_0030:  brtrue.s   IL_001f
+    IL_0032:  leave.s    IL_0045
+  }
+  finally
+  {
+    IL_0034:  ldloc.0
+    IL_0035:  isinst     ""System.IDisposable""
+    IL_003a:  stloc.3
+    IL_003b:  ldloc.3
+    IL_003c:  brfalse.s  IL_0044
+    IL_003e:  ldloc.3
+    IL_003f:  callvirt   ""void System.IDisposable.Dispose()""
+    IL_0044:  endfinally
+  }
+  IL_0045:  ret
+}");
+        }
+
+        [Fact]
+        public void TestConstantNullableWithGetEnumeratorPattern()
+        {
+            var source = @"
+using System;
+using System.Collections;
+public struct C
+{
+    public static void Main()
+    {
+        foreach (var i in (C?)null)
+        {
+            Console.Write(i);
+        }
+    }
+
+    public IEnumerator GetEnumerator() => throw null;
+}";
+            CompileAndVerify(source, parseOptions: TestOptions.RegularPreview)
+                .VerifyIL("C.Main", @"
+{
+  // Code size       64 (0x40)
+  .maxstack  2
+  .locals init (System.Collections.IEnumerator V_0,
+                C? V_1,
+                C V_2,
+                System.IDisposable V_3)
+  IL_0000:  ldloca.s   V_1
+  IL_0002:  dup
+  IL_0003:  initobj    ""C?""
+  IL_0009:  call       ""C C?.Value.get""
+  IL_000e:  stloc.2
+  IL_000f:  ldloca.s   V_2
+  IL_0011:  call       ""System.Collections.IEnumerator C.GetEnumerator()""
+  IL_0016:  stloc.0
+  .try
+  {
+    IL_0017:  br.s       IL_0024
+    IL_0019:  ldloc.0
+    IL_001a:  callvirt   ""object System.Collections.IEnumerator.Current.get""
+    IL_001f:  call       ""void System.Console.Write(object)""
+    IL_0024:  ldloc.0
+    IL_0025:  callvirt   ""bool System.Collections.IEnumerator.MoveNext()""
+    IL_002a:  brtrue.s   IL_0019
+    IL_002c:  leave.s    IL_003f
+  }
+  finally
+  {
+    IL_002e:  ldloc.0
+    IL_002f:  isinst     ""System.IDisposable""
+    IL_0034:  stloc.3
+    IL_0035:  ldloc.3
+    IL_0036:  brfalse.s  IL_003e
+    IL_0038:  ldloc.3
+    IL_0039:  callvirt   ""void System.IDisposable.Dispose()""
+    IL_003e:  endfinally
+  }
+  IL_003f:  ret
+}");
+        }
+
+        [Fact]
+        public void TestForeachNullLiteral()
+        {
+            var source = @"
+using System;
+public class C
+{
+    public static void Main()
+    {
+        foreach (var i in null)
+        {
+            Console.Write(i);
+        }
+    }
+}";
+            CreateCompilation(source, parseOptions: TestOptions.RegularPreview)
+                .VerifyDiagnostics(
+                    // (7,27): error CS0186: Use of null is not valid in this context
+                    //         foreach (var i in null)
+                    Diagnostic(ErrorCode.ERR_NullNotValid, "null").WithLocation(7, 27)
+                    );
+        }
+
+        [Fact]
+        public void TestForeachDefaultLiteral()
+        {
+            var source = @"
+using System;
+public class C
+{
+    public static void Main()
+    {
+        foreach (var i in default)
```

Comment:
>default [](start = 26, length = 7)

Are there tests for when `default(object)` has some extension defined?

Reply:
@YairHalberstadt I don't see any additional tests with regular `defaut(object)`, not the `default` literal.

---
In reply to: [443874353](https://github.com/dotnet/roslyn/pull/45160#discussion_r443874353) [](ancestors = 443874353)